social_core/backends/id4me.py

```diff
@@ -0,0 +1,270 @@
+
+    def get_or_create_state(self):
+        if self.STATE_PARAMETER or self.REDIRECT_STATE:
+            name = self.name + '_state'
+            state = self.strategy.session_get(name)
+            if state is None:
+                state = self.state_token()
+                self.strategy.session_set(name, state)
+        else:
+            state = None
+        return state
+
+    def get_scope(self):
+        scope = self.setting('SCOPE', {})
+        if not scope:
+            scope = self.DEFAULT_SCOPE
+        return scope
+
+    def get_scope_argument(self):
+        param = {'scope': 'openid'}
+        scope = self.get_scope()
+        if scope:
+            param['claims'] = json.dumps({'userinfo': scope})
+        return param
+
+    def oidc_config_authority(self):
+        return self.get_json('https://' + self.strategy.session_get(self.name + '_authority') +
+                             '/.well-known/openid-configuration')
+
+    def oidc_config_agent(self):
+        return self.get_json('https://' + self.strategy.session_get(self.name + '_agent') +
+                             '/.well-known/openid-configuration')
+
+    def authorization_url(self):
+        return self.oidc_config_authority().get('authorization_endpoint')
+
+    def access_token_url(self):
+        return self.oidc_config_authority().get('token_endpoint')
+
+    def id_token_issuer(self):
+        return [self.strategy.session_get(self.name + '_authority'),
+                'https://' + self.strategy.session_get(self.name + '_authority'),
+                self.strategy.session_get(self.name + '_authority').replace('https://', '')]
+
+    def userinfo_url(self):
+        return self.oidc_config_agent().get('userinfo_endpoint')
+
+    def jwks_uri(self):
+        return self.oidc_config_authority().get('jwks_uri')
+
+    def get_agent_keys(self):
+        return self.request(self.oidc_config_agent().get('jwks_uri')).json()['keys']
+
+    def get_jwks_keys(self):
+        keys = self.get_remote_jwks_keys()
+        return keys
+
+    def find_valid_key(self, id_token):
+        for key in self.get_jwks_keys():
+            header = jwt.get_unverified_header(id_token)
+            if header['kid'] == key['kid']:
+                if 'alg' not in key:
+                    key['alg'] = 'RS256' if key['kty'] == 'RSA' else 'ES256'
+                return key
+
+    def find_agent_valid_key(self, id_token):
+        for key in self.get_agent_keys():
+            header = jwt.get_unverified_header(id_token)
+            if header['kid'] == key['kid']:
+                if 'alg' not in key:
+                    key['alg'] = 'RS256' if key['kty'] == 'RSA' else 'ES256'
+                return key
+
+    def auth_complete(self, *args, **kwargs):
+        self.validate_state()
+        identity = self.strategy.session_get(self.name + '_identity')
+        openid_configuration = self.get_identity_record(identity)
+        if 'clp' not in openid_configuration:
```

Comment:
RP shall never need to read out 'clp', as the user-info delegation to Agent shall rely on distributed claims